src/server.rs

```diff
@@ -58,65 +54,567 @@ use protocol::{
+                                MissType::Normal => {
+                                    me.stats.borrow_mut().cache_misses += 1;
+                                }
+                                MissType::CacheReadError => {
+                                    me.stats.borrow_mut().cache_read_errors += 1;
+                                }
+                                MissType::ForcedRecache => {
+                                    me.stats.borrow_mut().cache_misses += 1;
+                                    me.stats.borrow_mut().forced_recaches += 1;
+                                }
+                            }
+                            me.stats.borrow_mut().cache_read_miss_duration += duration;
+                            cache_write = Some(future);
+                        }
+                        CompileResult::NotCacheable => {
+                            me.stats.borrow_mut().cache_misses += 1;
+                            me.stats.borrow_mut().non_cacheable_compilations += 1;
+                        }
+                        CompileResult::CompileFailed => {
+                            me.stats.borrow_mut().compile_fails += 1;
+                        }
+                    };
+                    let Output { status, stdout, stderr } = out;
+                    status.code()
+                        .map_or_else(
+                            || trace!("CompileFinished missing retcode"),
+                            |s| { trace!("CompileFinished retcode: {}", s); finish.set_retcode(s) });
+                    //TODO: sort out getting signal return on Unix
+                    finish.set_stdout(stdout);
+                    finish.set_stderr(stderr);
+                }
+                Err(_) => {
+                    me.stats.borrow_mut().cache_errors += 1;
+                    //TODO: figure out a better way to communicate this?
+                    finish.set_retcode(-2);
+                }
+            };
+            res.set_compile_finished(finish);
+            let send = tx.send(Ok(res));
+
+            let cache_write = cache_write.then(move |result| {
+                match result {
+                    Err(e) => {
+                        debug!("Error executing cache write: {}", e);
+                        me.stats.borrow_mut().cache_write_errors += 1;
+                    }
+                    //TODO: save cache stats!
+                    Ok(Some(info)) => {
+                        debug!("[{}]: Cache write finished in {}.{:03}s", info.object_file, info.duration.as_secs(), info.duration.subsec_nanos() / 1000_000);
+                        me.stats.borrow_mut().cache_writes += 1;
+                        me.stats.borrow_mut().cache_write_duration += info.duration;
+                    }
+
+                    Ok(None) => {}
+                }
+                Ok(())
+            });
+
+            send.join(cache_write).then(|_| Ok(()))
+        });
+
+        self.handle.spawn(task);
```

Comment:
Yeah this is where I felt like futures really started to shine, the overall flow ended up being much easier to follow afterwards.